src/io_threads.c

```diff
@@ -126,30 +289,390 @@ static void IOJobQueue_peek(const IOJobQueue *jq, job_handler *handler, void **d
+
+typedef struct delayedJob {
+    job_handler handler;
+    int slot;
+    char data[];
+} delayedJob;
+
+/* Global thread-local storage for delayed jobs */
+static __thread list *thread_delayed_jobs = NULL;
+
+/*
+ * executionContext
+ * -1: CTX_NONE (no execution context)
+ * -2: CTX_EXCLUSIVE (exclusive access to the whole database)
+ * >= 0: slot number (indicating we're operating on a specific slot)
+ */
+#define CTX_NONE -1
+#define CTX_EXCLUSIVE -2
+
+/* Global context tracking for deferred queue operations - mainthread only */
+static int dq_context = CTX_NONE;
+
+static deferredQueue *getDeferredQueue(int slot) {
+    if (slot == -1) {
+        return &deferredCmdExclusive;
+    } else {
+        return &slot_use_info[slot];
+    }
+}
+
+static int isClientListEmpty(deferredQueue *queue) {
+    return queue->pending_clients == NULL || listLength(queue->pending_clients) == 0;
+}
+
+static int isJobListEmpty(deferredQueue *queue) {
+    return queue->deferred_jobs == NULL || listLength(queue->deferred_jobs) == 0;
+}
+
+static int isDqEmpty(deferredQueue *queue) {
+    return isJobListEmpty(queue) && isClientListEmpty(queue);
+}
+
+/* Check if a deferred queue is available for immediate processing */
+static int dqAvailable(deferredQueue *queue) {
+    return queue->refcount == 0;
+}
+
+/* Returns the thread id handling the given slot, -1 if no thread is handling it */
+static int getSlotTid(int slot) {
+    if (dqAvailable(getDeferredQueue(slot))) return -1;
+    return getDeferredQueue(slot)->cur_tid;
+}
+
+static void setSlotTid(int slot, int tid) {
+    getDeferredQueue(slot)->cur_tid = tid;
+}
+
+/* Increment the reference count of a deferred queue */
+static void dqIncr(deferredQueue *queue) {
+    queue->refcount++;
+}
+
+/* Create a new job with the given handler and data */
+static listNode *createJobNode(int slot, job_handler handler, size_t data_size, void *data) {
+    /* Allocate memory for job structure plus data using flexible array member */
+    listNode *node = zmalloc(sizeof(listNode) + sizeof(delayedJob) + data_size);
+    delayedJob *job = (delayedJob *)(node + 1);
+    job->slot = slot;
+    job->handler = handler;
+    if (data_size) {
+        memcpy(job->data, data, data_size); /* Copy data to the flexible array member */
+    }
+    node->value = job;
+
+    return node;
+}
+
+/* Process a job immediately or add it to queue based on refcount */
+static void processOrAddJob(deferredQueue *q, listNode *jobNode) {
+    if (q->refcount == 0) {
+        delayedJob *job = listNodeValue(jobNode);
+        job->handler(job->data);
+        zfree(jobNode);
+    } else {
+        if (q->deferred_jobs == NULL) {
+            q->deferred_jobs = listCreate();
+        }
+        listLinkNodeTail(q->deferred_jobs, jobNode);
+    }
+}
+
+/* Returns whether the given command requires exclusive access to the whole database. */
+static int isDBExclusiveCmd(struct serverCommand *cmd, int slot) {
```

Comment:
```suggestion
static int isServerExclusiveCmd(struct serverCommand *cmd, int slot) {
```
These aren't really DB exclusive, since we don't allow those to be concurrently executed.

Reply:
Done